### DISCUSSION TITLE
Testing Kafka Connectors - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#309] [Announcements]
# Testing Kafka Connectors - Gunnar Morling

Kafka Connect is a key factor for the wide-spread adoption of Apache Kafka:
a framework and runtime environment for connectors,
it makes the task of getting data either into Kafka or out of Kafka solely a matter of configuration,
rather than a bespoke programming job.
There’s dozens, if not hundreds, of readymade source and sink connectors,
allowing you to create no-code data pipelines between all kinds of databases, APIs, and other systems.

There may be situations though where there is no existing connector matching your requirements,
in which case you can implement your own custom connector using the Kafka Connect framework.
Naturally, this raises the question of how to test such a Kafka connector,
making sure it propagates the data between the connected external system and Kafka correctly and completely.
In this blog post I’d like to focus on testing approaches for Kafka Connect source connectors,
i.e. connectors like Debezium, which ingest data from an external system into Kafka.
Very similar strategies can be employed for testing sink connectors, though.

https://www.morling.dev/blog/testing-kafka-connectors/

**Comment by bobmarks** [comment]
Great article! I used this as a basis for an integration test i was doing myself. However, I couldn't get your method of creating a JAR file to work (Kafka Connect logs didn't show any `Add plugin` entries). I eventually got it working my simply executing `./gradlew shadowJar` from the Java integration test. I posted the solution to https://stackoverflow.com/questions/77128033/testing-custom-kafka-connect-smt-using-test-containers if you are interested.

**Comment by asyncjason** [comment]
Yep, awesome article! Thanks, I really appreciate it. I spent today building an integration test for Debezium MySQL Source and JDBC Sink with cp-schema-registry and redpanda console. Will be very nice for debugging when I start some upcoming work I have to build some custom transformers. My first time with TestContainers. It's great! Cheers
